src/app/MessageDef/SubscribeRequest.cpp

```diff
@@ -0,0 +1,334 @@
+        {
+            err = CHIP_NO_ERROR;
+        }
+    }
+    ReturnLogErrorOnFailure(err);
+    return reader.ExitContainer(mOuterContainerType);
+}
+#endif // CHIP_CONFIG_IM_ENABLE_SCHEMA_CHECK
+
+CHIP_ERROR SubscribeRequest::Parser::GetAttributePathList(AttributePathList::Parser * const apAttributePathList) const
+{
+    TLV::TLVReader reader;
+    ReturnLogErrorOnFailure(mReader.FindElementWithTag(chip::TLV::ContextTag(kCsTag_AttributePathList), reader));
+    VerifyOrReturnLogError(chip::TLV::kTLVType_Array == reader.GetType(), CHIP_ERROR_WRONG_TLV_TYPE);
+    return apAttributePathList->Init(reader);
+}
+
+CHIP_ERROR SubscribeRequest::Parser::GetEventPathList(EventPathList::Parser * const apEventPathList) const
+{
+    TLV::TLVReader reader;
+    ReturnLogErrorOnFailure(mReader.FindElementWithTag(chip::TLV::ContextTag(kCsTag_EventPathList), reader));
+    VerifyOrReturnLogError(chip::TLV::kTLVType_Array == reader.GetType(), CHIP_ERROR_WRONG_TLV_TYPE);
+    return apEventPathList->Init(reader);
+}
+
+CHIP_ERROR
+SubscribeRequest::Parser::GetAttributeDataVersionList(AttributeDataVersionList::Parser * const apAttributeDataVersionList) const
+{
+    TLV::TLVReader reader;
+    ReturnLogErrorOnFailure(mReader.FindElementWithTag(chip::TLV::ContextTag(kCsTag_AttributeDataVersionList), reader));
+    VerifyOrReturnLogError(TLV::kTLVType_Array == reader.GetType(), CHIP_ERROR_WRONG_TLV_TYPE);
+    return apAttributeDataVersionList->Init(reader);
+}
+
+CHIP_ERROR SubscribeRequest::Parser::GetEventNumber(uint64_t * const apEventNumber) const
+{
+    return GetUnsignedInteger(kCsTag_EventNumber, apEventNumber);
+}
+
+CHIP_ERROR SubscribeRequest::Parser::GetMinIntervalSeconds(uint16_t * const apMinIntervalSeconds) const
+{
+    return GetUnsignedInteger(kCsTag_EventNumber, apMinIntervalSeconds);
+}
+
+CHIP_ERROR SubscribeRequest::Parser::GetMaxIntervalSeconds(uint16_t * const apMaxIntervalSeconds) const
+{
+    return GetUnsignedInteger(kCsTag_EventNumber, apMaxIntervalSeconds);
+}
+
+CHIP_ERROR SubscribeRequest::Parser::GetKeepExistingSubscriptions(bool * const apKeepExistingSubscription) const
+{
+    return GetSimpleValue(kCsTag_KeepExistingSubscriptions, chip::TLV::kTLVType_Boolean, apKeepExistingSubscription);
+}
+
+CHIP_ERROR SubscribeRequest::Parser::GetIsProxy(bool * const apIsProxy) const
+{
+    return GetSimpleValue(kCsTag_IsProxy, chip::TLV::kTLVType_Boolean, apIsProxy);
+}
+
+CHIP_ERROR SubscribeRequest::Builder::Init(chip::TLV::TLVWriter * const apWriter)
+{
+    return InitAnonymousStructure(apWriter);
+}
+
+AttributePathList::Builder & SubscribeRequest::Builder::CreateAttributePathListBuilder()
+{
+    // skip if error has already been set
+    if (mError == CHIP_NO_ERROR)
+    {
+        mError = mAttributePathListBuilder.Init(mpWriter, kCsTag_AttributePathList);
+        ChipLogFunctError(mError);
+    }
+    else
+    {
+        mAttributePathListBuilder.ResetError(mError);
```

Comment:
would it be the same and simpler to follow if we always reset?

```cpp
mAttributePathListBuilder.ResetError(mError);

if (mError == CHIP_NO_ERROR) {
    mError = mAttributePathListBuilder.Init(mpWriter, kCsTag_AttributePathList);
    ChipLogFunctError(mError);
}

return mAttributePathListBuilder;
```

or is this logically different (i.e. the path builder may contain an error and we do not want to clear it with the Request error)?

I find the logic of containing several separate errors a bit hard to follow. What is the invariant that we try to maintain regarding the relationship between mError and mAttributePathListBuilder::mError? What do combinations mean of "both no error", "both error", "one error, one no error"? Are the 2 variants of "one has error, one does not have error" even valid?

Reply:
Since this pattern is used throughout, we have errors for:

mError (data member)
mAttributePathListBuilder::mError
mEventPathListBuilder::mError
mAttributeDataVersionListBuilder::mError

so the combinations of having/not having errors seems to expand exponentially.
Still wondering what the invariant is - it seems to me we would want a single 'global error' instead of 4 separate ones. Is this correct? Could we desing our code to have such an approach?

Reply:
just remove mAttributePathListBuilder.ResetError(mError), there is no need to pass the error inside nested container
currently they are separete, and maintained for particular container, for example,
AttributeDataVersionList::Builder attributeDataVersionList = subscribeRequestBuilder.CreateAttributeDataVersionListBuilder();
NL_TEST_ASSERT(apSuite, subscribeRequestBuilder.GetError() == CHIP_NO_ERROR); //  the rror is inside top container
    aAttributeDataVersionListBuilder.AddVersion(1).aAttributeDataVersionListBuilder.EndOfAttributeDataVersionList();
    NL_TEST_ASSERT(apSuite, aAttributeDataVersionListBuilder.GetError() == CHIP_NO_ERROR);
===> the error is processed inside nested container.

I am still thinking of how to optimize the nested layer for better pattern

Reply:
https://github.com/project-chip/connectedhomeip/issues/8228